### DISCUSSION TITLE
How can I get my GH Action tests to use the NPM version of a peer package?

### DISCUSSION THREAD

**Discussion by slifty** [community/community#27035] [Actions]
<p>I have a yarn monorepo, and am making a change to one of the packages that <em>other</em> packages rely on.</p>
<p>I increment my package version on publication, so although my PR includes breaking changes, I have not adjusted the <code>package.json</code> version.  Unfortunately i seems that something in my GH actions (maybe something in <code>actions/checkout@v2</code>?) seems to have some logic that involves not checking out if the dependency matches the local version.</p>
<p>Is there any way to have it pull from NPM instead?</p>
<p>Here’s the action: <a href="https://github.com/tvkitchen/appliances/blob/aeb1455f166a7a5ad21e886ed69c494972f8372a/.github/workflows/test.yml" class="inline-onebox" rel="noopener nofollow ugc">appliances/test.yml at aeb1455f166a7a5ad21e886ed69c494972f8372a · tvkitchen/appliances · GitHub</a></p>

**Comment by shrink** [comment, marked as answer]
<aside class="quote no-group quote-modified" data-username="slifty" data-post="1" data-topic="188367">
<div class="title">
<div class="quote-controls"></div>
 slifty:</div>
<blockquote>
<p>Unfortunately GH actions <code>actions/checkout@v2</code> seems to have some logic that involves not checking out if the dependency matches the local version.</p>
</blockquote>
</aside>
<p>I don’t think that’s correct, certainly, I have not observed that before: I think this is an issue specific to the monorepo set-up you have. The monorepo tooling is responsible for creating the appropriate dependency linkage through <code>node_modules</code> – <code>actions/checkout</code> just pulls down your repository contents.</p>
<p>As I understand it, the expectation of a monorepo is that the packages are all compatible with each commit, that is, if you introduce a breaking change to Package A then you must update Package B <em>at the same time</em> to be compatible. A monorepo (in the node world at least) is not multiple independent packages in one repository, it’s multiple dependent packages in one repository.</p>
<p>A good first step to debugging this would be to upload an artifact containing the whole Action workspace, from there you can step through each dependency and identify any issues, e.g:</p>
<pre><code class="lang-yaml">- name: Upload runner Workspace as artifact
  if: always()
  uses: actions/upload-artifact@v2
  with:
      path: "."
      name: "workspace"
</code></pre>

**Comment by slifty** [comment]
<p>Much appreciated and you are certainly right – (as you were responding I actually modified that section to be less conclusion-jumping).  I don’t know why this didn’t manifest when I was using other CI solutions, but I think it was just luck.</p>
<p>I also appreciate your point about the monorepo expectations!  Given that, I’ll take this situation as actually a legitimately failing test and go fix my code.</p>

**Comment by shrink** [comment]
<aside class="quote no-group" data-username="slifty" data-post="3" data-topic="188367">
<div class="title">
<div class="quote-controls"></div>
 slifty:</div>
<blockquote>
<p>I also appreciate your point about the monorepo expectations! Given that, I’ll take this situation as actually a legitimately failing test and go fix my code.</p>
</blockquote>
</aside>
<p>I could be wrong: during my own monorepo set-up process, I struggled to find any authoritative description of exactly how I should expect some of the finer points to be addressed. I know that Yarn (and npm, and Lerna) <em>do</em> – as you describe in your opening post – pull from npm when they cannot resolve a version locally so if you do want to have backwards incompatibility within the monorepo, it is possible:</p>
<ol>
<li>Make a backwards-incompatible change in <code>package-a</code>
</li>
<li>Update <code>package-a/package.json</code>'s <code>version</code> to <code>2.0.0-alpha.1</code> (where <code>2</code> is the next major version)</li>
<li>Run <code>yarn install</code> and observe that any package in the monorepo depending on version 1.x.x of <code>package-a</code> will pull <code>package-a</code> from the registry instead of using the local copy</li>
</ol>
